### DISCUSSION TITLE
ccache may be too eager to remove object files? (Maybe some changes in the last 12 months?)

### DISCUSSION THREAD

**Discussion by zephyrus00jp** [ccache/ccache#292] [General]
I am just observing something that disturbed me in the least seven days.
I know cache behavior depends on the content of the hashed files, so it may not apply to everyone.

I develop patches for Mozilla Thunderbird (TB) on my local PC running linux (actually an image inside virtualbox, but that should not matter).

I use ccache for the compilation. I depend on it very much since I use mercurial queue (mqueue)  to manage local patches. With mqueue, I apply patches and then back them out, repeatedly while I develop patches to submit to mozilla. So I often end up source files that are touched/recreated with new timestamp, but the content is the same as before. Great opportunities for cache.

I am updaing ccache using GNU/Linux Debian's package and it is the latest version, I think.
I notice the following  behavior last week , but this undesirable behavior could have  been there for several months.
(I used to use my customized version of ccache which I created when I worked on the support of "-gsplit-dwarf" for a while, but then since the patch was included in ccache, I let Debia's packaging system to update ccache as new versions becomes available.)

What happened. 

During the RE-compilation of TB source tree AFTER a complete build of newly fetched TB source tree 
I saw recompilation of some (actually MANY)  source files during the second rebuild which should not happen using ccache: this has never happened before (except for a few files which seem to contain __DATE__ macro in them.).
[Well, the file layout of TB source tree changed quite a lot in the last three months since I fetched the TB source tree in mid-Aprill and so I was forced to shuffle the file layout, rewrite some scripts, and then recompile the new TB source I fetched at the end of July by fixing some issues along the way. Because of this total reshuffle, I was very attentive to find any abnormal behavior of build process when I notice the unnecessary compilation issue of ccache.],

I was so taken aback since ccache never behaved this way, and reading the log file, I found that the 
object files for the source files that forced re-compilation during the second build were removed in earlier compilation steps when the NUMBER of FILES in the hash  bucket was deemed too large.
ccache unlinked files from the hash bucket so that the total number of files went down below certain threshold.
[Here I am refering to the leaf subdirectory of the cache area when I say "hash bucket".]

So basically, to me, ccache was removing the object files that should have been kept and  used for later re-compilation process too prematurely.

OK, it may be that the hash function could have been channged, or the content of the files changed in a manner that make a larger number of files to fall in a hash bucket unfortunately. 
That is a unfortunate possibility, and I simply got unluckly, MAYBE.

But after using ccache for a few years, I was so surprised. Could this be a statistical flule?
I re-read the ccache manual carefully.

Now, I am curious if the removal of files per hash bucket in an LRU manner has been touched in the last 12 months.  If so, there may have been a subtle change that could cause a little too eager removal.
By reading the manual, I also notice the factor (default 0.8) when ccache culles the number of files so that the number of files do NOT go over certain number defined for per hash bucket. The manual states ccache removes the files until the number of files per hash bucket is reduced below 0.8 (default value) x the maximum file # per hash bucket.
Maybe this factor was NOT applied before in the earlier version of ccache (?)  (i.e., the number of files could remain up to the threshold before the removal took place.)

Actually, I was not aware that the removal was based on the number of files per hash bucket, but I had the impression that the removal was triggered based on the total size in the cache vs the cache size limit, or the total number of files in the cache as a whole, NOT the number of files per hash bucket. Maybe I was wrong.

Anyway, I relied upon ccache NOT to invoke unnecessary compilation and it served me well for the last few years in a somewhat cramped cacche area (I think the ccache used 3.5 GB when I set 4.0 GB or something like that for a long while.)
This tight space consumption by ccache was really nice.

After this recent unpleasant discovery, though,  I had to bump ccache limit to 6GB and now see only 5 files that get recompiled. This is much better. 

Still a mystery: two files seem to be either written with  __DATE__ macro in the source, or to include a header file that has a timestamp explicitly written it,  and so I can understand the recompilation. 
But other three files were NOT re-compiled before. Of course, Mozilla TB source files could have been modified to include such macros or unique ID per compilation  to force recompilation: I have not had the time to looki into it.

Enough tirade.

I am very grateful for the existence of ccacche, and I was even  motivitated to contribute the patches for "-gsplit-dwarf" because it is so useful.

If anyone had a similar experience of ccache suddenly removing objects in a stable config  in the last  12 months, and being forced to make the ccache size limit to a much larger limit to avoid such recompilation, I am interestined in hearing the story. 

To be honest, I feel there may be a sutble coding issue that removes files a bit too eagerly, but again, caching behavior depends on the content of the files (hash confclit, etc.), it is hard to say my experience is common. 
That is why I would like to hear if there are similar experiences out there. If so, I will be motivated to look into the source code again.
Otherwise, I probably had to be content with the BIG ccache size limit although ccache only use about half.

Thank you for sharing the great program, I cannot live without ccache on my local PC.

TIA

**Comment by afbjorklund** [comment]
Normally ccache will only clean on file size (default 5G), not on the number of files - although it is possible to configure if wanted... What is your ccache configuration and exact version ?

The other scenario with the mystery files, sounds like something that the hash debugging feature (#279) would be able to help with ? It should make it easier to diff the "before" and "after"

**Comment by zephyrus00jp** [comment]
After I decreased the ccache cache size to 6GB, and re-build the tree, I see these lines in the logfile:

[2018-08-08T13:51:33.355971 11753] Need to clean up /FF-NEW/cache-dir/3 since it holds 374460 KiB (limit: 366210 KiB)
[2018-08-08T13:58:21.524147 27355] Need to clean up /FF-NEW/cache-dir/3 since it holds 366364 KiB (limit: 366210 KiB)

So, it looks that it was NOT the # of files, maybe I misread, but the size of the files in this particular group of hash buckets (no. 3).
Hmm. This is unfortunate. Is there a way to bump up the size limit for the top-level hash group?

**Comment by zephyrus00jp** [comment]
Sorry, I meant to post the following message first.

Too bad, I re-created the cache from scratch and
I thought doing so would be best done by removing the cache directory completely and
re-build MS TB source.
Thus I lost the logfiles from the last week. Hmm...

Anyway, this is the |ccache -s| output after I actually increased the cache size to 7 GB (after my original post)

OK, I have no idea why, but the object file size seem to have become
much larger than before. Ugh...

Note 5.6 GB size in the output. So could this be the cause of forced
removal of object files?  But, if so, then the message in the log file
about the # of files exceeding limit per bucket is misleading, and
also the number of files seem to be excessive. See Note at the end.

Despite the appearance in the output of |ccache -s|,  /etc/ccache.conf DOES NOT exist.

++ ccache -s
cache directory                     /FF-NEW/cache-dir
primary config                      /FF-NEW/cache-dir/ccache.conf
secondary config      (readonly)    /etc/ccache.conf
stats zero time                     Sat Aug  4 08:26:46 2018
cache hit (direct)                 16608
cache hit (preprocessed)             549
cache miss                          7846
cache hit rate                     68.62 %
called for link                      732
called for preprocessing             250
compile failed                        76
preprocessor error                   617
cache file missing                     9
bad compiler arguments                 4
autoconf compile/link                652
no input file                         60
cleanups performed                    41
files in cache                     17302
cache size                           5.6 GB
max cache size                       7.0 GB
ccache -c --max-size=7.0GB
++ ccache -c --max-size=7.0GB
Cleaned cache
Set cache size limit to 7.0 GB


max_size = 7.0GB

--

Note: the number of files per hash bucket.

I really think the new hash function and the handling of files in hash
bucket is/was fishy.

> files in cache                     17302

When I noticed the recompilation issue, I recall that the log message
stated my cache exceeded the per bucket limit of approximately 3,000
and that is why ccache was removing cached files. But in the re-built
cache from scratch above shows only 17302 files.  How can evenly
distributed cache could contain a bucket that exceeded 3K file limit
!? (Of course, the message could have been buggy, and indeed the size
limit due to the bloated object files in the new source tree was to
blame for the unexpected deletion of object files.)

OR does the bucket mean not the leaf subdirectory, but the
intermediate directory that has 16 leaf subdirectories?

PS: as in the previous message, it could be the volume of files in a particular hash bucket group that trigger the unlinking of files.




I will try reducing the cache size incrementally if I can obtain some
hints.  (But the ccache cache may have a group of history-sensitive
remains from previous runs, I am not sure if I can re-create the
originally buggy condition very well...)  Also, I freed up the partition
where the ccache cache lies, and so the tight file system condition may
also play role, but that won't be in effect, in this new test
environment.

**Comment by zephyrus00jp** [comment]
ccache version is:

ccache --version
ccache version 3.4.2

I have a feeling that this automatic cleanup was not reliable in some earlier version of ccache. But it gets run very reliably.

Since it seems a bucket group goes over the size limit when cache size is set to 6GB,
and 
export CCACHE_LIMIT_MULTIPLE=1.2
does not help (I suppose multiple is capped at 1.0 internally),
I am experimenting CCACHE_COMPRESS=true to see if I can avoid with the total cache size of 6.0GB.
I think I can get away with the unnecessary recompilation by setting it 7.0GB.

I am not sure what changed in the last 12 months, but it seems that gcc 7 produces larger debug information file (.dwo) than gcc 6 and this size increase seems to have caused the overflowing over the size/16  limit per top-level hash bucket group.

> The other scenario with the mystery files, sounds like something that the hash debugging feature (#279) would be able to help with ? It should make it easier to diff the "before" and "after"
I wonder if this worked under older ccache version.
If so, I could certainly compare the behavior of the
older ccache + gcc-6 and the latest ccache + gcc-7.

It is nice to know that automatic removal now works very reliably but
given the somewhat unbalanced distribution of file sizes in the hash bucket group in real development scenario, it would be great if we have a mechanism to relax the per-top-level-bucket-group size limit so that a few top-level bucket group may go over the per-group limit, but the whole cache size is not exceeded.

Mozilla's build system concatenate source files under a source directory together carefully into a few gigantic source files, and compile them instead of visiting each file one by one to save time to invoke compiler. (I believe Apple's development tool pursued the pre-compiled header approach rather eagerly, but mozilla developers opted for this approach.)
So a single removal of an object file means an invocation of compiler against the gigantic concatenated  source file, and so avoiding a single unnecessary compilation [i.e., the file is not touched] if possible is a sure win.

 TIA

**Comment by afbjorklund** [comment]
At the moment, the number of buckets is hardcoded to 16 (since it is based on a hex digit). I don't think that md4 gives any guarantees about even distribution.

Compression sounds like a good idea.

**Comment by zephyrus00jp** [comment]
Thank you for the comment.

My study results have to wait until Tuesday. I am on a summer vacation from today and go to my home town. As soon as I come back, I will report my finding in detail.

**Comment by afbjorklund** [comment]
I think the suggestions here (only do one cleanup) is quite the opposite of #168 (do 256, instead of 16)

So it might have to be changed into an option, depending on what feature is considered more important:

- bucket distribution and "fairness" of cleanup

- lockfile congestion and runtime performance

Another option is disabling the automatic cleanup altogether, and run it manually in a controlled way ?

This is needed for things like NFS, where it is very easy to deny service by triggering multiple cleanups...

**Comment by jrosdahl** [comment]
@zephyrus00jp: I have trouble following your tirade since you seem to have been confused by several things, but perhaps you're less confused now? :slightly_smiling_face:

Just to clarify:

* Like @afbjorklund already mentioned, the default is to only consider cache size when deciding when to clean up, not the number of files in the cache.
* ccache 3.4.2 has an important bug fix related to cleanup. From [ccache 3.4.2 release notes](https://ccache.samba.org/releasenotes.html#_ccache_3_4_2):
    > The cleanup algorithm has been fixed to not misbehave when files are removed by another process while the cleanup process is running. Previously, too many files could be removed from the cache if multiple cleanup processes were triggered at the same time, in extreme cases trimming the cache to a much smaller size than the configured limits.

  In other words: Before 3.4.2, the number of files in different hash buckets could vary significantly due to a race condition when cleaning up. With 3.4.2, the hash buckets should be relatively similar in size. The expected difference is up to 20% since the default limit multiple is 0.8. If they differ with more than that, then there might be more bugs lurking.

@zephyrus00jp:
> export CCACHE_LIMIT_MULTIPLE=1.2
> does not help (I suppose multiple is capped at 1.0 internally),

Yes, it's capped at 1.0 since higher values make no sense ("I want to keep 120% of the cache when cleaning up!!!!1"). And the only thing that will come out of limit multiple 1.0 is very poor performance since cleanup will be triggered for more or less every cache miss, and a cleanup run has an overhead cost that is way higher than single cache misses.

@zephyrus00jp:
> it would be great if we have a mechanism to relax the per-top-level-bucket-group size limit so that a few top-level bucket group may go over the per-group limit, but the whole cache size is not exceeded.

ccache deliberately restricts its view to just one of the sixteen hash buckets for performance reasons. It would be costly to have to consider all hash buckets on a cache miss.

@afbjorklund:
> I don't think that md4 gives any guarantees about even distribution.

Even though MD4 is severly broken for cryptographic purposes, it still has the properties of a cryptographic hash function, one of which is that it's very uniformly distributed. I don't have any good reference to back up my claim at the moment, but I would be very surprised if it's the hash function that makes hash buckets differently sized.

**Comment by jrosdahl** [comment]
@zephyrus00jp: Closing this now, please comment if you're still wondering anything.
